### PR TITLE
Add support for new video_remand_hearing move_type

### PR DESCRIPTION
**Pull request by smoothcontract**
### Jira link

P4-1816

### What?

- [x] Added new `video_remand_hearing` to `move_type` enum
- [x] Allow video remand hearing moves to be created without specifying a `to_location`
- [x] Enforce validation of `police` type for `from_location` when creating `video_remand_hearing`
- [x] Add new `MoveTypeValidator` class

### Why?

- This is the first of many new move types to be added, each introducing additional validation rules on the permitted from and to location types. As the first step to implementing support for these rules, validation of move type and associated location types is extracted to a new custom validator; this allows the logic to be shared across the move model and move runner.

### Have you? (optional)

- [x] Updated API docs if necessary - this will be updated further in future PR's to document the behaviour of the different move types

### Deployment risks (optional)

- Adds support for new move type, but does not change behaviour of existing move types

### CODE REVIEW

File: app/services/moves/move_type_validator.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    def includes?(move_types)
+      move_types.include?(record.move_type.to_s)
+    end
+
+    def human_move_type
+      record.move_type.to_s.humanize(capitalize: false)
+    end
+
+    def validate_police_from_location
+      record.errors.add(:from_location, "must be a police location for #{human_move_type}") unless record.from_location&.police?
```
Comment (alexdesi) [inline app/services/moves/move_type_validator.rb:26]: maybe I am wrong, when we add an error, we should specify
- the field,
- the type of error, for example :too_long and 
- the message, 
here the type is missing 🤔

File: swagger/v1/move.yaml [hunk 2]
```diff
@@ -44,6 +44,7 @@ Move:
           - prison_recall
           - prison_transfer
           - police_transfer
+          - video_remand_hearing
```
Comment (willfish) [inline swagger/v1/move.yaml:47]: Presumably we want this on v2 moves, too?

**Comment by willfish** [comment]
I foresee that validator becoming complex and horrible ☠️

File: app/services/moves/move_type_validator.rb (hunk 1 above)
Comment (smoothcontract) [reply app/services/moves/move_type_validator.rb:26]: Not quite - the second argument can be a string or a symbol to be resolved in the locale file. We have plenty of examples of similar things, but I'll add an explicit test to check the returned error message text.

File: swagger/v1/move.yaml (hunk 2 above)
Comment (smoothcontract) [reply swagger/v1/move.yaml:47]: GPWM - I'll update the v2 docs too 👍

**Comment by smoothcontract** [comment]
> I foresee that validator becoming complex and horrible ☠️ 🔮

You could well be right, but so far I'm only seeing the need to validate against from and to locations, so at least it's all in one place. We can erect suitable warning signs and barbed wire fences as needed.

File: app/services/move_events/params_validator.rb [hunk 3]
```diff
@@ -32,17 +32,21 @@ class ParamsValidator
...
     end
 
-    validates_with LocationValidator, locations: [:from_location], if: -> { type == 'lockouts' }
-    validates_with LocationValidator, locations: [:to_location], if: -> { type == 'redirects' }
+    validates_with LocationValidator, locations: [:from_location_id], if: -> { type == 'lockouts' }
+    validates_with LocationValidator, locations: [:to_location_id], if: -> { type == 'redirects' }
+    validates_with Moves::MoveTypeValidator, if: -> { type == 'redirects' }
 
-    def initialize(params)
+    delegate :move_type, :from_location, to: :move
```
Comment (martyn-w) [inline app/services/move_events/params_validator.rb:39]: I think these names could potentially be confusing:
```
from_location --> move's from_location
from_location_id --> event's from_location
```

how about:
```
move_from_location --> move's from_location
event_from_location --> event's from_location(_id)
```
?

**Comment by martyn-w** [comment]
Looks good so far.
1. I think the naming of `from_location` vs `from_location_id` is confusing in the move event; consider a different name instead. Sorry I didn't pick up the full implication of this earlier.
2. Add a test for move_type_validator - its a bit fidley as you'll need a new class, see e.g. spec/services/location_validator_spec.rb

File: app/services/move_events/params_validator.rb (hunk 3 above)
Comment (smoothcontract) [reply app/services/move_events/params_validator.rb:39]: Discussed offline, this isn't trivial to change as the `MoveTypeValidator` needs to believe that this is really a `Move` model with the same `move_type` and `from_location` attributes. Agreed to have another look at that in the next PR for the other upcoming move types. In the meantime this is enough to unblock front end work.

**Pull request merged**